### PR TITLE
Publish raw mjpeg stream directly via compressed image topic

### PR DESCRIPTION
**Pull request by boitumeloruf**
Hi @flynneva,

I have another PR. At this point, however, I prefer it as a feature idea and suggestion for a possible implementation of the same. I am not too happy with my implementation yet (especially in usb_cam_node.cpp), thus, I would instead use it as a basis for discussion.

Since most of the USB cameras directly provide a MJPEG stream, it would be most efficient not to let the driver do the decoding but rather publish it via the compressed image topic (this is also suggested [here](https://wiki.ros.org/compressed_image_transport#Publishing_compressed_images_directly)). In this way, any node that uses the image_transport plugin can directly connect to the MJPEG stream. Furthermore, if the camera is connected to an edge device the decoding can be transferred to a more powerful device while at the same time having the benefit of an efficient data transfer using MJPEG.

What are your thoughts on this? My implementation allows selecting a raw MJPEG stream by setting the pixel_format in the config YAML to 'mjpeg'. I have tested it successfully by subscribing to '/image_raw/compressed' using rqt_image_view, as well as a custom subscription node.

### CODE REVIEW

File: include/usb_cam/formats/mjpeg.hpp
Comment (flynneva) [inline include/usb_cam/formats/mjpeg.hpp:74]: @boitumeloruf this switch should be moved to its own function. input is the avDeviceFormat, output would be the corresponding ROS format. Should probably split it into two separate functions - one for the ROS format, the other for the channels.

Something like `get_ros_format_from_av_format(const AVPixelFormat & avDeviceFormat)` and then same thing for the channels.

**Comment by flynneva** [comment]
@boitumeloruf apologies this took so long for me to get to but I like this idea a lot. If youre up for coming back to this I'd be happy to merge it after rebasing / the one suggestion I had above. We can iterate / improve on it once its merged

**Comment by boitumeloruf** [comment]
Hello @flynneva, Yes, I am happy to implement your suggestions. :)

**Comment by boitumeloruf** [comment]
Hi @flynneva, I have implemented your suggestions. I also fixed some build issues and code style issues. Let me know if there is stuff left to do.

File: include/usb_cam/formats/av_pixel_format_helper.hpp
Comment (flynneva) [inline include/usb_cam/formats/av_pixel_format_helper.hpp:975]: @boitumeloruf I know its not necessary, but could you move the `default` cases for the switches here to be last? You could also remove the `break` then from each of the default cases

File: include/usb_cam/formats/av_pixel_format_helper.hpp
Comment (flynneva) [inline include/usb_cam/formats/av_pixel_format_helper.hpp:981]: @boitumeloruf writing each case like this will make the code much more read-able I think (and a lot less clutter)

```suggestion
    case AVPixelFormat::AV_PIX_FMT_RGB24: ros_format = usb_cam::constants::RGB8; break;
```

**Review by flynneva — APPROVED** [review]
@boitumeloruf I only could see two minor suggestions / nits so I'll go ahead and approve and merge this. We can tackle those comments in another PR if we really want to 👍🏼

**Comment by boitumeloruf** [comment]
Great, Thanks!

**Comment by wendwosenbb** [comment]
@boitumeloruf 

would you elaborate how to use this feature?  on how to configure  pixel_format and av_device_format.  I also asked about this on issue #346 

my goal is to reduce cpu usage with this new feature you provided. The following information about the hardware

  my camera is Imx291 usb camera. 

The device formats supports:

`v4l2-ctl --list-formats-ex --device /dev/windshield_camera

`

ioctl: VIDIOC_ENUM_FMT
Type: Video Capture

    [0]: 'MJPG' (Motion-JPEG, compressed)
            Size: Discrete 1920x1080
                    Interval: Discrete 0.033s (30.000 fps)
                    Interval: Discrete 0.040s (25.000 fps)
                    Interval: Discrete 0.050s (20.000 fps)
                    Interval: Discrete 0.067s (15.000 fps)
                    Interval: Discrete 0.100s (10.000 fps)
            Size: Discrete 1280x720
                    Interval: Discrete 0.033s (30.000 fps)
                    Interval: Discrete 0.040s (25.000 fps)
                    Interval: Discrete 0.050s (20.000 fps)
                    Interval: Discrete 0.067s (15.000 fps)
                    Interval: Discrete 0.100s (10.000 fps)
            Size: Discrete 800x600
                    Interval: Discrete 0.033s (30.000 fps)
                    Interval: Discrete 0.050s (20.000 fps)
                    Interval: Discrete 0.067s (15.000 fps)
                    Interval: Discrete 0.100s (10.000 fps)
                    Interval: Discrete 0.200s (5.000 fps)
            Size: Discrete 640x480
                    Interval: Discrete 0.033s (30.000 fps)
                    Interval: Discrete 0.040s (25.000 fps)
                    Interval: Discrete 0.050s (20.000 fps)
                    Interval: Discrete 0.067s (15.000 fps)
                    Interval: Discrete 0.100s (10.000 fps)
            Size: Discrete 320x240
                    Interval: Discrete 0.033s (30.000 fps)
                    Interval: Discrete 0.040s (25.000 fps)
                    Interval: Discrete 0.050s (20.000 fps)
                    Interval: Discrete 0.067s (15.000 fps)
                    Interval: Discrete 0.100s (10.000 fps)
    [1]: 'YUYV' (YUYV 4:2:2)
            Size: Discrete 1280x720
                    Interval: Discrete 0.100s (10.000 fps)
            Size: Discrete 1920x1080
                    Interval: Discrete 0.200s (5.000 fps)
            Size: Discrete 800x600
                    Interval: Discrete 0.100s (10.000 fps)
                    Interval: Discrete 0.200s (5.000 fps)
            Size: Discrete 640x480
                    Interval: Discrete 0.033s (30.000 fps)
                    Interval: Discrete 0.040s (25.000 fps)
                    Interval: Discrete 0.050s (20.000 fps)
                    Interval: Discrete 0.067s (15.000 fps)
                    Interval: Discrete 0.100s (10.000 fps)
            Size: Discrete 320x240
                    Interval: Discrete 0.033s (30.000 fps)
                    Interval: Discrete 0.040s (25.000 fps)
                    Interval: Discrete 0.050s (20.000 fps)
                    Interval: Discrete 0.067s (15.000 fps)
                    Interval: Discrete 0.100s (10.000 fps)`
`

The usb_cam driver is showing me pixel formats supported

[usb_cam_node_exe-2] This driver supports the following formats:
[usb_cam_node_exe-2] rgb8
[usb_cam_node_exe-2] yuyv
[usb_cam_node_exe-2] yuyv2rgb
[usb_cam_node_exe-2] uyvy
[usb_cam_node_exe-2] uyvy2rgb
[usb_cam_node_exe-2] mono8
[usb_cam_node_exe-2] mono16
[usb_cam_node_exe-2] y102mono8
[usb_cam_node_exe-2] raw_mjpeg
[usb_cam_node_exe-2] mjpeg2rgb
[usb_cam_node_exe-2] m4202rgb

**Pull request merged**
